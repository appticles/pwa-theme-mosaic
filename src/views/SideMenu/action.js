import config from 'react-global-configuration';

export const REQUEST_PAGES = 'REQUEST_PAGES';
export const RECEIVE_PAGES = 'RECEIVE_PAGES';

export const requestPages = () => ({
  type: REQUEST_PAGES,
});

export const receivePages = pages => ({
  type: RECEIVE_PAGES,
  pages,
});

export const fetchPages = (params = {}) => (dispatch) => {
  dispatch(requestPages());

  let url;
  if (params && params.id) {
    url = config.get('export').pages + '/' + String(params.id);
  } else {
    url =
      config.get('export').pages +
      '?' +
      Object.keys(params)
        .map(k => k + '=' + encodeURIComponent(params[k]))
        .join('&');
  }
  return fetch(url)
    .then(response => response.json())
    .then(json => dispatch(receivePages(json)))
    .catch(() => {
      dispatch(receivePages([]));
    });
};
