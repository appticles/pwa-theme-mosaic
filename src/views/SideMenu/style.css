.sidebar.menu-pages .ui.header {
  position: static;
  color: #999;
  font-size: 23px;
  font-weight: 100;
  line-height: 45px;
  font-family: "Roboto Condensed", sans-serif;
  margin: 1rem;
  display: flex;
  align-items: flex-start;
}

.sidebar.menu-pages .ui.header .icon {
  margin-left: 0;
  width: 1rem;
}

/* LIST OVERRIDE */
.sidebar.menu-pages .ui.divided.relaxed.list {
  height: unset;
  margin: 0 1rem;
}

.sidebar.menu-pages .ui.divided.relaxed.list .item {
  display: flex;
  align-items: center;
  position: relative;
}

.sidebar.menu-pages .ui.divided.relaxed.list .item div.header {
  font-family: "Roboto Condensed";
  font-size: 18px;
  font-weight: normal;
  line-height: 1.5rem;
  color: #000000 !important;
  padding: 0.5rem;
  text-align: left;
}

.sidebar.menu-pages .ui.divided.relaxed.list .item > i:nth-child(3) {
  position: absolute;
  right: 0;
}

/* MODAL */
/* .ui.scrolling.modal .header {
  border: none;
}
.modal .image {
  margin: auto;
} */

.ui.modal .content {
  position: relative;
}
