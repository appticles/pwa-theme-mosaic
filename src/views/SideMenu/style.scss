@import "../../utilities/_variables.scss";

.ui.menu-pages.sidebar.visible {
  border-color: $border-color;
  i {
    color: $buttons-icon-color;
  }
  .list {
    padding: $base-spacing-xsm;

    .header {
      font-family: $base-font-family, sans-serif;
      color: $base-text-color;
      margin-top: $base-spacing-sm;
    }

    .item {
      display: flex;
      align-items: center;
      height: 8vh;
      position: relative;
      border-color: $border-color;

      .content {
        display: flex;
        align-items: center;

        h3 {
          font-family: $base-font-family, sans-serif;
          color: $base-text-color;
        }
      }

      .icon.triangle {
        position: absolute;
        right: 0;
      }
    }
  }
}
