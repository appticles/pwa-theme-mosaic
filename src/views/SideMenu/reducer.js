import { combineReducers } from 'redux';
import _ from 'lodash';
import PropTypes from 'prop-types';
import { REQUEST_PAGES, RECEIVE_PAGES } from './action';

export const pagePropType = PropTypes.shape({
  id: PropTypes.number.isRequired,
  parent: PropTypes.number.isRequired,
  title: PropTypes.shape({
    rendered: PropTypes.string.isRequired,
  }).isRequired,
  slug: PropTypes.string.isRequired,
  content: PropTypes.shape({
    rendered: PropTypes.string.isRequired,
  }),
});

const items = (state = [], action) => {
  switch (action.type) {
  case REQUEST_PAGES:
    return state;
  case RECEIVE_PAGES:
    if (Array.isArray(action.pages)) {
      return _.unionBy(state, action.pages, 'id');
    }
    return _.unionBy(state, [action.pages], 'id');
  default:
    return state;
  }
};

export const getPages = state => state.items;
export default combineReducers({ items });
