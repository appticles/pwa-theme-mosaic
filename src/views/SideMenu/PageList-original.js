import React, { Component } from 'react';
import { Link } from 'react-router-dom';
import { List, Header, Icon, Modal, Image, Transition } from 'semantic-ui-react';

import './style.css';

export default class PageList extends Component {
  constructor(props) {
    super(props);
    this.state = { parentIds: [], pages: [], parentPageTitle: ['Go to'], goBack: false, visible: true };
  }

  componentWillReceiveProps() {
    this.setState({ parentIds: [...this.state.parentIds, 0] });
    // resets state when menu reopens
    this.props.visible === true && this.setState({ parentIds: [], pages: [], parentPageTitle: ['Go to'], goBack: false });
  }
  setParentsList(parentId, pages, parentPageTitle) {
    this.setState({
      parentIds: [...this.state.parentIds, parentId],
      pages: [...this.state.pages, pages],
      parentPageTitle: [...this.state.parentPageTitle, parentPageTitle],
      goBack: false,
      visible: !this.state.visible,
    });
  }
  setParentsBackList() {
    // copie + setstate
    this.state.parentIds.pop();
    const poppedParentIds = this.state.parentIds;
    this.state.parentPageTitle.pop();
    const poppedPageTitles = this.state.parentPageTitle;
    this.setState({
      parentId: poppedParentIds,
      parentsPageTitle: poppedPageTitles,
      goBack: true,
      visible: !this.state.visible,
    });
  }

  render() {
    const checkParent = this.state.parentIds[this.state.parentIds.length - 1];
    let parents;
    // iF?
    this.state.goBack === false ? (parents = this.props.pages.filter(page => page.parent === checkParent)) : (parents = this.state.pages.pop());

    function getImage(sourceImg) {
      let imageSource;
      if (sourceImg) {
        imageSource = sourceImg[0].source_url;
        return imageSource;
      }
      return imageSource;
    }

    return (
      <Transition animation="pulse" duration={700} visible={this.state.visible}>
        <List divided relaxed>
          {checkParent === 0 || parents.length === 0 ? (
            <Header>{this.state.parentPageTitle[this.state.parentPageTitle.length - 1]}</Header>
          ) : (
            <Header>
              <Icon name="caret left" onClick={() => this.setParentsBackList()} />
              {this.state.parentPageTitle[this.state.parentPageTitle.length - 1]}
            </Header>
          )}
          {parents.map(page => (
            <List.Item key={Math.random()}>
              <List.Icon name="linkify" size="large" verticalAlign="middle" />
              <List.Content>
                <Modal closeIcon trigger={<List.Header as="a">{page.title.rendered}</List.Header>}>
                  <Modal.Header dangerouslySetInnerHTML={{ __html: page.title.rendered }} />
                  {/* eslint */}
                  <Image src={getImage(page._embedded['wp:featuredmedia'])} size="medium" />
                  <Modal.Content dangerouslySetInnerHTML={{ __html: page.content.rendered }} />
                  <div className="go-to">
                    <Link to={'/page/' + page.slug + '/' + page.id}>More</Link>
                  </div>
                </Modal>
              </List.Content>
              {page.children.length > 0 ? (
                <List.Icon
                  link
                  name="triangle right"
                  size="large"
                  verticalAlign="middle"
                  onClick={() => this.setParentsList(page.id, parents, page.title.rendered)}
                />
              ) : (
                ''
              )}
            </List.Item>
          ))}
        </List>
      </Transition>
    );
  }
}
