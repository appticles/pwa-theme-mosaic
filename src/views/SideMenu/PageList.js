import React, { Component } from 'react';
import { Link } from 'react-router-dom';
import { List, Header, Icon, Modal, Image } from 'semantic-ui-react';

import './style.css';

export default class PageList extends Component {
  state = { parentIds: [], pages: [], parentPageTitle: ['Go to'], goBack: false };

  componentWillReceiveProps() {
    this.setState({ parentIds: [...this.state.parentIds, 0] });
    // resets state when menu reopens
    this.props.visible === true && this.setState({ parentIds: [], pages: [], parentPageTitle: ['Go to'], goBack: false });
  }
  setParentsList(parentId, pages, parentPageTitle) {
    this.setState({
      parentIds: [...this.state.parentIds, parentId],
      pages: [...this.state.pages, pages],
      parentPageTitle: [...this.state.parentPageTitle, parentPageTitle],
      goBack: false,
    });
  }
  setParentsBackList() {
    this.state.parentIds.pop();
    const poppedParentIds = this.state.parentIds;
    this.state.parentPageTitle.pop();
    const poppedPageTitles = this.state.parentPageTitle;
    this.setState({ parentId: poppedParentIds, parentsPageTitle: poppedPageTitles, goBack: true });
  }

  render() {
    const checkParent = this.state.parentIds[this.state.parentIds.length - 1];
    let parents;
    this.state.goBack === false ? (parents = this.props.pages.filter(page => page.parent === checkParent)) : (parents = this.state.pages.pop());

    function getImage(sourceImg) {
      let imageSource;
      if (sourceImg) {
        imageSource = sourceImg[0].source_url;
        return imageSource;
      }
      return imageSource;
    }

    return (
      <List divided relaxed>
        {checkParent === 0 ? (
          <Header>{this.state.parentPageTitle[this.state.parentPageTitle.length - 1]}</Header>
        ) : (
          <Header>
            <Icon name="caret left" onClick={e => this.setParentsBackList()} />
            {this.state.parentPageTitle[this.state.parentPageTitle.length - 1]}
          </Header>
        )}
        {parents.map(page => (
          <List.Item key={Math.random()}>
            <List.Icon name="linkify" size="large" verticalAlign="middle" />
            <List.Content>
              <Modal closeIcon trigger={<List.Header as="a">{page.title.rendered}</List.Header>}>
                <Modal.Header dangerouslySetInnerHTML={{ __html: page.title.rendered }} />
                <Image src={getImage(page._embedded['wp:featuredmedia'])} size="medium" />
                <Modal.Content dangerouslySetInnerHTML={{ __html: page.content.rendered }} />
                <div className="go-to">
                  <Link to={'/page/' + page.slug + '/' + page.id}>More</Link>
                </div>
              </Modal>
            </List.Content>
            {page.children.length > 0 ? (
              <List.Icon
                link
                name="triangle right"
                size="large"
                verticalAlign="middle"
                onClick={e => this.setParentsList(page.id, parents, page.title.rendered)}
              />
            ) : (
              ''
            )}
          </List.Item>
        ))}
      </List>
    );
  }
}
