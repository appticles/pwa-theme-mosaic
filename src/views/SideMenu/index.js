import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';
import { Sidebar } from 'semantic-ui-react';

import { fetchPages } from './action';
import { getPages, pagePropType } from './reducer';
import PageList from './PageList';

import './style.css';

class SideMenu extends Component {
  componentWillMount() {
    const { dispatch } = this.props;
    dispatch(fetchPages({}));
  }

  makePagesList() {
    const { pages } = this.props;
    pages.map((page) => {
      page.children = [];
      return page;
    });
    function placeChildren(currentLevel, nextLevel) {
      currentLevel.map((parent) => {
        nextLevel.map((child) => {
          if (child.parent === parent.id) {
            parent.children.push(child);
          }
          return nextLevel;
        });
        return currentLevel;
      });
    }
    placeChildren(pages, pages);
  }

  render() {
    const { pages } = this.props;
    this.makePagesList();
    return (
      <Sidebar visible={this.props.isVisible} direction="right">
        <PageList pages={pages} />
      </Sidebar>
    );
  }
}
SideMenu.propTypes = {
  dispatch: PropTypes.func.isRequired,
  isVisible: PropTypes.bool.isRequired,
  pages: PropTypes.arrayOf(pagePropType).isRequired,
};
const mapStateToProps = state => ({
  pages: getPages(state.pages),
});
function mapDispatchToProps(dispatch) {
  return Object.assign({ dispatch }, bindActionCreators({ fetchPages }, dispatch));
}
export default connect(mapStateToProps, mapDispatchToProps)(SideMenu);
