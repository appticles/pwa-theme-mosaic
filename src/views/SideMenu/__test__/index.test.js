import React from 'react';
import renderer from 'react-test-renderer';

import mockStore from '../../../__mocks__/redux-mock-store';

import SideMenu from '../index';

jest.mock('../PageList', () => 'Page list test');

describe('Side menu', () => {
  it('renders side menu', () => {
    const store = mockStore({
      pages: {
        items: [],
        isFetching: 1,
      },
      sideMenuVisible: true,
    });
    const loading = 1;
    expect(renderer.create(<SideMenu store={store} loading={loading} fetchPages={jest.fn()} />)).toMatchSnapshot();
  });
  it('renders hidden side menu', () => {
    const store = mockStore({
      pages: {
        items: [],
        isFetching: 0,
      },
      sideMenuVisible: true,
    });
    const loading = 1;
    expect(renderer.create(<SideMenu store={store} loading={loading} fetchPages={jest.fn()} />)).toMatchSnapshot();
  });
  it('renders side menu', () => {
    const store = mockStore({
      pages: {
        items: [
          {
            id: 53,
            parent: 5,
            title: {
              rendered: 'Page Title',
            },
            slug: 'page-slug',
            content: {
              rendered: 'Page content',
            },
            link: 'page-link',
            author: 68,
            comment_status: 'open',
            featured_media: 36,
            status: 'status',
          },
        ],
        isFetching: 0,
      },
      sideMenuVisible: true,
    });
    const loading = 1;
    expect(renderer.create(<SideMenu store={store} loading={loading} fetchPages={jest.fn()} />)).toMatchSnapshot();
  });
});
