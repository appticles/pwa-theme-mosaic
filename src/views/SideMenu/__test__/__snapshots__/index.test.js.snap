// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Side menu renders hidden side menu 1`] = `
<div
  className="ui ui push right visible sidebar menu-pages menu"
>
  
  <Page list test
    pages={Array []}
    text={undefined}
  />
</div>
`;

exports[`Side menu renders side menu 1`] = `
<div
  className="ui ui push right visible sidebar menu-pages menu"
>
  <div
    className="ui active loader"
  />
  <Page list test
    pages={Array []}
    text={undefined}
  />
</div>
`;

exports[`Side menu renders side menu 2`] = `
<div
  className="ui ui push right visible sidebar menu-pages menu"
>
  
  <Page list test
    pages={
      Array [
        Object {
          "author": 68,
          "comment_status": "open",
          "content": Object {
            "rendered": "Page content",
          },
          "featured_media": 36,
          "id": 53,
          "link": "page-link",
          "parent": 5,
          "slug": "page-slug",
          "status": "status",
          "title": Object {
            "rendered": "Page Title",
          },
        },
      ]
    }
    text="Visit website"
  />
</div>
`;
